kernel/kernel-api/src/main/java/io/delta/kernel/Transaction.java

```diff
@@ -0,0 +1,152 @@
+     * @param dataActions  Iterable of data actions to commit. These data actions are generated by
+     *                    the {@link Transaction#generateAppendActions(TableClient, Row,
+     *                    CloseableIterator, DataWriteContext)}. The {@link CloseableIterable}
+     *                    allows the Kernel to access the list of actions multiple times (in case of
+     *                    retries to resolve the conflicts due to other writers to the table).
+     *                    Kernel provides a in-memory based implementation of
+     *                    {@link CloseableIterable} with utility API
+     *                    {@link CloseableIterable#inMemoryIterable(CloseableIterator)}
+     * @return {@link TransactionCommitResult} status of the successful transaction.
+     * @throws ConcurrentWriteException when the transaction has encountered a non-retryable
+     *                                  conflicts or exceeded the maximum number of retries reached.
+     *                                  The connector needs to rerun the query on top of the latest
+     *                                  table state and retry the transaction.
+     */
+    TransactionCommitResult commit(TableClient tableClient, CloseableIterable<Row> dataActions)
+            throws ConcurrentWriteException;
+
+    /**
+     * Given the logical data that needs to be written to the table, convert it into the required
+     * physical data depending upon the table Delta protocol and features enabled on the table.
+     * Kernel takes care of adding any additional column or removing existing columns that doesn't
+     * need to be in physical data files. All these transformations are driven by the Delta protocol
+     * and table features enabled on the table.
+     * <p>
+     * The given data should belong to exactly one partition. It is the job of the connector to do
+     * partitioning of the data before calling the API. Partition values are provided as map of
+     * column name to partition value (as {@link Literal}). If the table is an un-partitioned table,
+     * then map should be empty.
+     *
+     * @param tableClient      {@link TableClient} instance to use.
+     * @param transactionState The transaction state
+     * @param dataIter         Iterator of logical data to transform to physical data. All the data
+     *                         in this iterator should belong to one physical partition.
+     * @param partitionValues  The partition values for the data. If the table is un-partitioned,
+     *                         the map should be empty
+     * @return Iterator of physical data to write to the data files.
+     */
+    static CloseableIterator<FilteredColumnarBatch> transformLogicalData(
+            TableClient tableClient,
+            Row transactionState,
+            CloseableIterator<FilteredColumnarBatch> dataIter,
+            Map<String, Literal> partitionValues) {
+        validatePartitionValues(
```

Comment:
Why are you validating the partition values _and_ throwing an unsupported operation exception?

Reply:
This line isn't removed as separate this PR from the larger PR. This is fixed now.